### ISSUE TITLE
Add support for gitstatusd

### ISSUE THREAD

**Issue by arkq** [nojhan/liquidprompt#742]
When git repository is very big (lots of submodules and submodules of submodules) running `git status` is very slow. Recently I've discovered a project https://github.com/romkatv/gitstatus which claims to be 10x baster than `git status`. And in fact it is (tested by myself on ~20GB repo, with many many number of files and submodules).

Unfortunately the `gitstatusd` is not a simple replacement for `git status`. It can be run as a daemon process which can be used for querying git status of given repository, so direct integration in liquidpromp would be required.

### Description

It would be nice to be able to switch from `git status` to `gitstatusd` if desired (or maybe automatically switch to `gitstatusd` if executable is available).

### How will this be useful?

It will improved responsiveness of liquidprompt when browsing git repository (especially big ones).

### Example prompt

The prompt should stay as it is right now.

**Comment by Rycieos** [comment]
This is for sure something we need to support. Likely we would want a config option for it (`LP_ENABLE_GITSTATUSD`), and autodetect if `gitstatus_check()` returns true, otherwise disable the feature. That would require the user to manage the sourcing of `gitstatus.plugin.sh` or `gitstatus.plugin.zsh` and running the `gitstatus_start()` function themselves.

The bigger trick will be the backend work inside `liquidprompt` to enable different ways to get the `git` data. Probably more work caching things.

**Comment by Rycieos** [comment]
No, no one is working on this to my knowledge. I took a shallow look at it before, and it seems doable, but a lot of work in liquidprompt internals to support multiple ways of fetching git data.

Honestly, this inspired me to think about other ways to speed up the prompt. Not to follow in @dolmen's shoes, but I have been toying around with the idea of a terminal prompt engine that runs as a daemon (like his [angel-PS1](https://github.com/dolmen/angel-PS1)). What kills most Liquid Prompt competitors is startup cost in loading the binary (which Liquid Prompt avoids because it runs native in the shell). Anyway, this is a tangent (that I would be happy to discuss with anyone interested).

All that to say that I have little interest in implementing this myself (at this time), as I would rather daemonize the whole prompt engine. So unless someone else picks this up, it will probably miss v2.2.

**Comment by Rycieos** [comment]
Surely you didn't actually delete 15 million lines of code, right? That is humorous.

I don't currently have interest in my other project I mentioned in this thread, so I think it's time I took a stab at this. It's time for a release in this project anyway.

**Comment by ottok** [comment]
The slowness is not related to how many files are deleted, but how much changes there are, build artifacts etc.

**Comment by Rycieos** [comment]
I implemented this on my `feature/gitstatus` branch. It looks like a bunch of tests failed (probably unset errors), but the basic functionality is working. I briefly tested it, and will continue to test it at work tomorrow. I would appreciate it if someone else could test it before I merge it.

The important docs can be found here: https://github.com/liquidprompt/liquidprompt/blob/3a9412fb5c5fa6d67726997d259fb172bc928aad/docs/config.rst?plain=1#L591-L612

**Comment by arkq** [comment]
> The issue with that is gitstatus_start has multiple options that configure how deep it searches.

Yes, I thought that maybe there should be a liquidprompt configuration option for `gitstatus_start` arguments, otherwise run it with default setup (no args). However, `PROMPT_COMMAND` might also be OK.

**Comment by Rycieos** [comment]
I forgot to mention previously (though I did add it to the docs, see [here](https://liquidprompt.readthedocs.io/en/latest/config.html#LP_ENABLE_GITSTATUSD)). gitstatusd does not support changed line data, so if you do not disable that with [`LP_ENABLE_VCS_LINES=0`](https://liquidprompt.readthedocs.io/en/latest/config.html#LP_ENABLE_VCS_LINES), the performance benefit will be limited.

**Comment by ottok** [comment]
Thanks for adding support for [gitstatus](https://github.com/romkatv/gitstatus)! Unfortunately it seems it is no longer maintained and does not support eg sha256 hashes. In https://github.com/romkatv/gitstatus/issues/411 the author says they will try to make their own shell prompt [powerlevel10k](https://github.com/romkatv/powerlevel10k) work without gitstatus in the future. If feels like a slightly risky new extra dependency to use gitstatus, although it does the job.

Any idea why upstream git hasn't adopted the improved status algorigth done in gitstatus?

**Comment by Rycieos** [comment]
> Unfortunately it seems it is no longer maintained and does not support eg sha256 hashes.

This is correct. Unfortunately not much to be done about that, unless someone takes up maintenance or a different project is started.

> Any idea why upstream git hasn't adopted the improved status algorigth done in gitstatus?

Because the improvements in gitstatusd are not in a better algorithm, but in its daemon nature. Since it runs constantly, it can skip loading the `git` program from disk, skip a `fork` call, and skips scanning the whole git directory every time (it caches some of it).

**Comment by ottok** [comment]
I did some more testing, and discovered that actually running `time git status --porcelain -b` yields just 0.2 seconds in a project, where the Liquid Prompt rendering takes ~5 seconds. So I went through and tested manually each git command I saw in https://github.com/liquidprompt/liquidprompt/blob/master/liquidprompt and found out that this is these are the slowes ones:

```
__lp_git_diff_shortstat_uncommitted()
$ time git diff --shortstat HEAD --
 13381 files changed, 81 insertions(+), 15104214 deletions(-)

real	0m2.188s
user	0m1.887s
sys	0m0.397s

__lp_git_diff_shortstat_unstaged()
$ time git diff --shortstat
 13381 files changed, 81 insertions(+), 15104214 deletions(-)

real	0m2.299s
user	0m1.930s
sys	0m0.456s
```

Thus I did this modification for myself locally:

```diff
diff --git a/liquidprompt.bak b/liquidprompt
index 6375804..50d9554 100644
--- a/liquidprompt.bak
+++ b/liquidprompt
@@ -2056,9 +2056,13 @@ __lp_git_diff_shortstat_lines() {  # diff_shortstat
 }
 
 __lp_git_diff_shortstat_uncommitted() {
-    if [[ -z ${_lp_git_diff_shortstat_uncommitted-} ]]; then
-        _lp_git_diff_shortstat_uncommitted="$(LC_ALL=C \git diff --shortstat HEAD -- 2>/dev/null)"
+  if [[ -z ${_lp_git_diff_shortstat_uncommitted-} ]]; then
+    if _lp_git_diff_shortstat_uncommitted="$(timeout 0.25 bash -c 'LC_ALL=C \git diff --shortstat HEAD -- 2>/dev/null')"; then
+      :
+    else
+      _lp_git_diff_shortstat_uncommitted=""
     fi
+  fi
 }
 
 # Get the number of changed files compared to HEAD.
@@ -2083,9 +2087,13 @@ _lp_git_uncommitted_lines() {
 }
 
 __lp_git_diff_shortstat_unstaged() {
-    if [[ -z ${_lp_git_diff_shortstat_unstaged-} ]]; then
-        _lp_git_diff_shortstat_unstaged="$(LC_ALL=C \git diff --shortstat 2>/dev/null)"
+  if [[ -z ${_lp_git_diff_shortstat_unstaged-} ]]; then
+    if _lp_git_diff_shortstat_unstaged="$(timeout 0.25 bash -c 'LC_ALL=C \git diff --shortstat 2>/dev/null')"; then
+      :
+    else
+      _lp_git_diff_shortstat_unstaged=""
     fi
+  fi
 }
```

Yes, the prompt won't have the number of files or lines modified for large projects, but at least the prompt is usable in the sense that it shows and is responsive.
